File: Platforms/Windows/Reign_Video_D3D9_Component/BufferLayout.cpp
#include "pch.h"
#include "BufferLayout.h"

namespace Reign_Video_D3D9_Component
{
	#pragma region Constructors
	BufferLayoutErrors BufferLayoutCom::Init(VideoCom^ video, BufferLayoutDescCom^ bufferLayoutDesc)
	{
		null();
		this->video = video;

		IDirect3DVertexDeclaration9* layoutTEMP;
		if (FAILED(video->device->CreateVertexDeclaration(bufferLayoutDesc->desc, &layoutTEMP)))
		{
			return BufferLayoutErrors::VertexDeclaration;
		}
		layout = layoutTEMP;

		return BufferLayoutErrors::None;
	}

	BufferLayoutCom::~BufferLayoutCom()
	{
		if (layout) layout->Release();
		null();
	}

	void BufferLayoutCom::null()
	{
		layout = 0;
	}
	#pragma endregion

	#pragma region Methods
	void BufferLayoutCom::Enable()
	{
		video->device->SetVertexDeclaration(layout);
	}
	#pragma endregion
}

File: Platforms/Win32/Reign.Video.D3D9/Buffers/BufferLayout.cs
using System;
using Reign.Video;
using Reign_Video_D3D9_Component;
using Reign.Core;

namespace Reign.Video.D3D9
{
	public class BufferLayout : Disposable, BufferLayoutI
	{
		#region Properties
		private BufferLayoutCom com;
		#endregion

		#region Constructors
		public static BufferLayout New(DisposableI parent, ShaderI shader, BufferLayoutDescI desc)
		{
			return new BufferLayout(parent, shader, desc);
		}

		public BufferLayout(DisposableI parent, ShaderI shader, BufferLayoutDescI desc)
		: base(parent)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				com = new BufferLayoutCom();
				var error = com.Init(video.com, ((BufferLayoutDesc)desc).com);

				switch (error)
				{
					case BufferLayoutErrors.VertexDeclaration: Debug.ThrowError("BufferLayout", "Failed to create vertex declaration"); break;
				}
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Enable()
		{
			com.Enable();
		}
		#endregion
	}
}
